server/launcher.go

```diff
@@ -18,47 +18,142 @@
 package server
 
 import (
-	"io"
 	"net"
 	"net/http"
 	"time"
 
+	"github.com/CanonicalLtd/UCWifiConnect/utils"
+
+	"fmt"
 	"log"
 )
 
+const (
+	address = ":8080"
+)
+
+// Server running state
+const (
+	Stopped RunningState = 0 + iota
+	Starting
+	Running
+	Stopping
+)
+
+// RunningState enum defining which server is up and running
+type RunningState int
+
+// State holds current server state
+var State = Stopped
+
+var listener net.Listener
+var done chan bool
+
 type tcpKeepAliveListener struct {
 	*net.TCPListener
 }
 
+// WaitForState waits for server reach certaing state
+func WaitForState(state RunningState) bool {
+	retries := 10
+	idle := 10 * time.Millisecond
+	for ; retries > 0 && State != state; retries-- {
+		time.Sleep(idle)
+		idle *= 2
+	}
+	return State == state
+}
+
 // Accept accepts incoming tcp connections
-func (ln tcpKeepAliveListener) Accept() (c net.Conn, err error) {
+func (ln tcpKeepAliveListener) Accept() (net.Conn, error) {
 	tc, err := ln.AcceptTCP()
 	if err != nil {
-		return
+		return tc, err
 	}
 	tc.SetKeepAlive(true)
 	tc.SetKeepAlivePeriod(3 * time.Minute)
 	return tc, nil
 }
 
-func listenAndServe(addr string, handler http.Handler) (sc io.Closer, err error) {
+func listenAndServe(addr string, handler http.Handler) error {
 
-	var listener net.Listener
+	if State != Stopped {
+		return fmt.Errorf("Server is not in proper stopped state before trying to start it")
```

Comment:
can you please change the error text throughout to use this format: "== wifi-connect/server: MSG". It makes it so much easier to find the log messages. (I missed the server when I cleaned up log messages a while back...)